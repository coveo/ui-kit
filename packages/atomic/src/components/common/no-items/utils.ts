import {i18n} from 'i18next';

export const getSummary = (
  i18n: i18n,
  query: string,
  hasResults: boolean,
  i18nKey: 'no-products' | 'no-results'
) => {
  if (hasResults) {
    return '';
  }

  const labelFor = `${i18nKey}-for`;

  return query
    ? i18n.t(labelFor, {
        interpolation: {escapeValue: false},
        query,
      })
    : i18n.t(i18nKey);
};
