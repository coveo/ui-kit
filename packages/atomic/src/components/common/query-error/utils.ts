import {i18n} from 'i18next';
import {KnownErrorType} from './known-error-types';

export const getErrorTitleFromErrorType = (
  i18n: i18n,
  organizationId: string,
  errorType?: string
) => {
  switch (errorType as KnownErrorType) {
    case 'Disconnected':
      return i18n.t('disconnected');

    case 'NoEndpointsException':
      return i18n.t('no-endpoints', {org: organizationId});
    case 'InvalidTokenException':
      return i18n.t('cannot-access', {org: organizationId});
    case 'OrganizationIsPausedException':
      return i18n.t('organization-is-paused', {org: organizationId});
    default:
      return i18n.t('something-went-wrong');
  }
};

export const getErrorDescriptionFromErrorType = (
  i18n: i18n,
  organizationId: string,
  url: string,
  errorType?: string
) => {
  switch (errorType as KnownErrorType) {
    case 'Disconnected':
      return i18n.t('check-your-connection', {url});
    case 'NoEndpointsException':
      return i18n.t('add-sources');
    case 'InvalidTokenException':
      return i18n.t('invalid-token');
    case 'OrganizationIsPausedException':
      return i18n.t('organization-will-resume', {
        org: organizationId,
      });
    default:
      return i18n.t('if-problem-persists');
  }
};

export const getAriaMessageFromErrorType = (
  i18n: i18n,
  organizationId: string,
  platformUrl: string,
  errorType?: string
) => {
  return `${getErrorTitleFromErrorType(i18n, organizationId, errorType)} ${getErrorDescriptionFromErrorType(i18n, organizationId, platformUrl, errorType)}`;
};
