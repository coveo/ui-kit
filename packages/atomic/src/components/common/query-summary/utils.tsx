import {FunctionalComponent, h} from '@stencil/core';
import {i18n} from 'i18next';

interface i18nKeyProps {
  first: number;
  last: number;
  total: number;
  query: string;
  isLoading: boolean;
  i18n: i18n;
}

const WrapHighlight: FunctionalComponent<{part?: string}> = (
  props,
  children
) => {
  return (
    <span
      class="font-bold"
      part={`highlight${props.part ? ` ${props.part}` : ''}`}
    >
      {children}
    </span>
  );
};

export const getQuerySummaryI18nParameters = ({
  first,
  last,
  query,
  total,
  isLoading,
  i18n,
}: i18nKeyProps) => {
  const i18nKey =
    query !== '' ? 'showing-results-of-with-query' : 'showing-results-of';

  const params = {
    first: first.toLocaleString(),
    last: last.toLocaleString(),
    total: total.toLocaleString(),
    query: query,
    count: last,
  };

  const highlights = {
    first: <WrapHighlight>{params.first}</WrapHighlight>,
    last: <WrapHighlight>{params.last}</WrapHighlight>,
    total: <WrapHighlight>{params.total}</WrapHighlight>,
    query: <WrapHighlight part="query">{params.query}</WrapHighlight>,
  };

  const ariaLiveMessage = isLoading
    ? i18n.t('loading-results')
    : i18n.t(i18nKey, params);

  return {i18nKey, highlights, ariaLiveMessage};
};
