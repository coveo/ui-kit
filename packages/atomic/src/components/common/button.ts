import {FunctionalComponentWithChildren} from '@/src/utils/functional-component-utils';
import {html} from 'lit';
import {ifDefined} from 'lit/directives/if-defined.js';
import {when} from 'lit/directives/when.js';
import {createRipple} from '../../utils/ripple';
import {
  getRippleColorForButtonStyle,
  getClassNameForButtonStyle,
  ButtonStyle,
} from './button-style';

export interface ButtonProps {
  style: ButtonStyle;
  onClick?(event?: MouseEvent): void;
  class?: string;
  text?: string;
  part?: string;
  type?: 'button' | 'submit' | 'reset' | 'menu';
  form?: string;
  role?:
    | 'status'
    | 'application'
    | 'checkbox'
    | 'button'
    | 'dialog'
    | 'img'
    | 'radiogroup'
    | 'toolbar'
    | 'listitem'
    | 'list'
    | 'separator';
  disabled?: boolean;
  ariaLabel?: string;
  ariaExpanded?: 'true' | 'false';
  ariaPressed?: 'true' | 'false' | 'mixed';
  ariaChecked?: 'true' | 'false' | 'mixed';
  ariaCurrent?: 'page' | 'false';
  ariaControls?: string;
  ariaHidden?: 'true' | 'false';
  tabIndex?: number;
  title?: string;
}

export const button: FunctionalComponentWithChildren<ButtonProps> = ({
  props,
  children,
}) => {
  const rippleColor = getRippleColorForButtonStyle(props.style);
  const className = getClassNameForButtonStyle(props.style);

  return html`<button
    type=${ifDefined(props.type)}
    title=${ifDefined(props.title)}
    tabindex=${ifDefined(props.tabIndex)}
    role=${ifDefined(props.role)}
    part=${ifDefined(props.part)}
    form=${ifDefined(props.form)}
    class=${props.class ? `${className} ${props.class}` : className}
    aria-pressed=${ifDefined(props.ariaPressed)}
    aria-label=${ifDefined(props.ariaLabel)}
    aria-hidden=${ifDefined(props.ariaHidden)}
    aria-expanded=${ifDefined(props.ariaExpanded)}
    aria-current=${ifDefined(props.ariaCurrent)}
    aria-controls=${ifDefined(props.ariaControls)}
    aria-checked=${ifDefined(props.ariaChecked)}
    @mousedown=${(e: MouseEvent) => createRipple(e, {color: rippleColor})}
    @click=${props.onClick}
    ?disabled=${props.disabled}
  >
    ${when(props.text, () => html`<span class="truncate">${props.text}</span>`)}
    ${children}
  </button>`;
};
