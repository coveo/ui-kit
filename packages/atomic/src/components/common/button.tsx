import {FunctionalComponent, h} from '@stencil/core';
import {createRipple} from '../../utils/ripple';

interface ButtonProps {
  style:
    | 'primary'
    | 'outline-primary'
    | 'outline-neutral'
    | 'outline-bg-neutral'
    | 'text-primary'
    | 'text-neutral'
    | 'text-transparent';
  onClick?(): void;
  class?: string;
  text?: string;
  part?: string;
  type?: string;
  role?: string;
  disabled?: boolean;
  ariaLabel?: string;
  ariaExpanded?: string;
  ariaPressed?: string;
  ariaChecked?: string;
  ariaCurrent?: string;
  title?: string;
  ref?(element?: HTMLButtonElement): void;
}

export const Button: FunctionalComponent<ButtonProps> = (props, children) => {
  let rippleColor: string;
  let className: string;
  switch (props.style) {
    case 'primary':
      className = 'btn-primary';
      rippleColor = 'primary';
      break;
    case 'outline-primary':
      className = 'btn-outline-primary';
      rippleColor = 'neutral';
      break;
    case 'outline-neutral':
      className = 'btn-outline-neutral';
      rippleColor = 'neutral';
      break;
    case 'outline-bg-neutral':
      className = 'btn-outline-bg-neutral';
      rippleColor = 'neutral';
      break;
    case 'text-primary':
      className = 'btn-text-primary';
      rippleColor = 'neutral';
      break;
    case 'text-neutral':
      className = 'btn-text-neutral';
      rippleColor = 'neutral';
      break;
    case 'text-transparent':
      className = 'btn-text-transparent';
      rippleColor = 'neutral-light';
      break;
  }

  const attributes = {
    class: props.class ? `${className} ${props.class}` : className,
    part: props.part,
    onClick: props.onClick,
    title: props.title,
    type: props.type,
    role: props.role,
    'aria-label': props.ariaLabel,
    'aria-expanded': props.ariaExpanded,
    'aria-pressed': props.ariaPressed,
    'aria-checked': props.ariaChecked,
    'aria-current': props.ariaCurrent,
    disabled: props.disabled,
    ref: props.ref,
  };

  return (
    <button
      {...attributes}
      onMouseDown={(e) => createRipple(e, {color: rippleColor})}
    >
      {props.text ? <span class="truncate">{props.text}</span> : null}
      {children}
    </button>
  );
};
