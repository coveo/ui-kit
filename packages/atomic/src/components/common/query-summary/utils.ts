import {FunctionalComponent} from '@/src/utils/functional-component-utils';
import {i18n} from 'i18next';
import {html} from 'lit';

interface i18nKeyProps {
  first: number;
  last: number;
  total: number;
  query: string;
  isLoading: boolean;
  i18n: i18n;
}

interface QuerySummaryLabels {
  loadingStatus: string;
  itemsForQuery: string;
  allItems: string;
}

const renderWrapHighlight: FunctionalComponent<{
  part?: string;
  content: string;
}> = ({props}) => {
  const part = `highlight${props.part ? ` ${props.part}` : ''}`;
  return html` <span class="front-bold" part="${part}">${props.content}</span>`;
};

export const getQuerySummaryI18nParameters = (props: i18nKeyProps) => {
  return getQuerySummaryData(props, {
    loadingStatus: 'loading-results',
    itemsForQuery: 'showing-results-of-with-query',
    allItems: 'showing-results-of',
  });
};

export const getProductQuerySummaryI18nParameters = (props: i18nKeyProps) => {
  return getQuerySummaryData(props, {
    loadingStatus: 'loading-products',
    itemsForQuery: 'showing-products-of-with-query',
    allItems: 'showing-products-of',
  });
};

const getQuerySummaryData = (
  {first, last, query, total, isLoading, i18n}: i18nKeyProps,
  {allItems, itemsForQuery, loadingStatus}: QuerySummaryLabels
) => {
  const i18nKey = query === '' ? allItems : itemsForQuery;

  const locale = i18n.language;
  const params = {
    first: first.toLocaleString(locale),
    last: last.toLocaleString(locale),
    total: total.toLocaleString(locale),
    query: query,
    count: last,
  };

  const highlights = {
    first: renderWrapHighlight({props: {content: params.first}}),
    last: renderWrapHighlight({props: {content: params.last}}),
    total: renderWrapHighlight({props: {content: params.total}}),
    query: renderWrapHighlight({props: {part: 'query', content: params.query}}),
  };

  const ariaLiveMessage = isLoading
    ? i18n.t(loadingStatus)
    : i18n.t(i18nKey, params);

  return {i18nKey, highlights, ariaLiveMessage};
};
