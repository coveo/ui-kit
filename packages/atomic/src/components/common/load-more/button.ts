import {i18n} from 'i18next';
import {button, ButtonProps} from '../button';

interface LoadMoreButtonProps {
  i18n: i18n;
  onClick: () => void;
  moreAvailable: boolean;
  label?: 'load-more-results' | 'load-more-products';
}

export function loadMoreButton({
  i18n,
  onClick,
  moreAvailable,
  label,
}: LoadMoreButtonProps) {
  if (!moreAvailable) {
    return;
  }
  const props: ButtonProps = {
    style: 'primary',
    part: 'load-more-results-button',
    class: 'my-2 p-3 font-bold',
    onClick: () => onClick(),
  };
  return button({props, children: i18n.t(label || 'load-more-results')});
}
