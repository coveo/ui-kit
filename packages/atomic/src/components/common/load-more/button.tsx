import {FunctionalComponent, h} from '@stencil/core';
import {i18n} from 'i18next';
import {Button} from '../button';

interface LoadMoreButtonProps {
  i18n: i18n;
  onClick: () => void;
  moreAvailable: boolean;
}

export const LoadMoreButton: FunctionalComponent<LoadMoreButtonProps> = ({
  i18n,
  onClick,
  moreAvailable,
}) => {
  if (!moreAvailable) {
    return;
  }
  return (
    <Button
      style="primary"
      text={i18n.t('load-more-results')}
      part="load-more-results-button"
      class="font-bold my-2 p-3"
      onClick={() => onClick()}
    ></Button>
  );
};
