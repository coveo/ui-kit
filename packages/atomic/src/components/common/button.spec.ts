import {createRipple} from '@/src/utils/ripple';
import {fireEvent, within} from '@storybook/test';
import {html, render} from 'lit';
import {vi} from 'vitest';
import {button, ButtonProps} from './button';

vi.mock('@/src/utils/ripple', () => ({
  createRipple: vi.fn(),
}));

describe('button', () => {
  let container: HTMLElement;

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    document.body.removeChild(container);
  });

  const renderButton = (props: Partial<ButtonProps>): HTMLButtonElement => {
    render(
      html`${button({
        props: {
          ...props,
          style: props.style ?? 'primary',
        },
        children: html``,
      })}`,
      container
    );
    return within(container).getByRole('button') as HTMLButtonElement;
  };

  it('should render a button in the document', () => {
    const props = {};
    const button = renderButton(props);
    expect(button).toBeInTheDocument();
  });

  it('should render a button with the correct style', () => {
    const props: Partial<ButtonProps> = {
      style: 'outline-error',
    };

    const button = renderButton(props);

    expect(button).toHaveClass('btn-outline-error');
  });

  it('should render a button with the correct text', () => {
    const props = {
      text: 'Click me',
    };

    const button = renderButton(props);

    expect(button.querySelector('span')?.textContent).toBe('Click me');
  });

  it('should wrap the button text with a truncate class', () => {
    const props = {
      text: 'Click me',
    };

    const button = renderButton(props);

    expect(button.querySelector('span')).toHaveClass('truncate');
  });

  it('should handle click event', async () => {
    const handleClick = vi.fn();
    const props = {
      onClick: handleClick,
    };

    const button = renderButton(props);

    await fireEvent.click(button);

    expect(handleClick).toHaveBeenCalled();
  });

  it('should apply disabled attribute', () => {
    const props = {
      disabled: true,
    };

    const button = renderButton(props);

    expect(button.hasAttribute('disabled')).toBe(true);
  });

  it('should apply aria attributes', () => {
    const props: Partial<ButtonProps> = {
      ariaLabel: 'button',
      ariaPressed: 'true',
    };

    const button = renderButton(props);

    expect(button.getAttribute('aria-label')).toBe('button');
    expect(button.getAttribute('aria-pressed')).toBe('true');
  });

  it('should apply custom class', () => {
    const props = {
      class: 'custom-class',
    };

    const button = renderButton(props);

    expect(button).toHaveClass('custom-class');
    expect(button).toHaveClass('btn-primary');
  });

  it('should apply part attribute', () => {
    const props = {
      part: 'button-part',
    };

    const button = renderButton(props);

    expect(button.getAttribute('part')).toBe('button-part');
  });

  it('should apply title attribute', () => {
    const props = {
      title: 'Button Title',
    };

    const button = renderButton(props);

    expect(button.getAttribute('title')).toBe('Button Title');
  });

  it('should apply tabindex attribute', () => {
    const props = {
      tabIndex: 1,
    };

    const button = renderButton(props);

    expect(button.getAttribute('tabindex')).toBe('1');
  });

  it('should apply role attribute', () => {
    const props: Partial<ButtonProps> = {
      role: 'button',
    };

    const button = renderButton(props);

    expect(button.getAttribute('role')).toBe('button');
  });

  it('should call onMouseDown when the mousedown event is fired on the button', async () => {
    const props: Partial<ButtonProps> = {};
    const button = renderButton(props);
    await fireEvent.mouseDown(button);
    expect(createRipple).toHaveBeenCalled();
  });

  it('should apply form attribute', () => {
    const props = {
      form: 'form-id',
    };

    const button = renderButton(props);

    expect(button.getAttribute('form')).toBe('form-id');
  });

  it('should apply type attribute', () => {
    const props: Partial<ButtonProps> = {
      type: 'submit',
    };

    const button = renderButton(props);

    expect(button.getAttribute('type')).toBe('submit');
  });
});
