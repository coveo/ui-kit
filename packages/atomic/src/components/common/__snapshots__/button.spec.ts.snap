// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`button > matches known snapshot 1`] = `
<button
  class="btn-primary"
>
  
      
  <span
    class="truncate"
  >
    Default test text
  </span>
  
      
  
    
</button>
`;
